### PR TITLE
vvc_deblock.c: add checkasm cases for strong, one-side, weak, mixed, shift

### PR DESCRIPTION
**Pull request by stone-d-chen**
Add cases by iteratively setting max_len_{p, q} and shift.

randomize_chroma_buffers() now takes both buffers and a parameter gen_strong. It will also internally handle buffer copying.

### CODE REVIEW

**Comment by stone-d-chen** [comment]
New tests

Note: Bench numbers aren't accurate.

```
AVX:
 - vvc_deblock.chroma [OK]
checkasm: all 24 tests passed
vvc_h_loop_filter_chroma_8_mix_no-shift_c: 60.1
vvc_h_loop_filter_chroma_8_mix_no-shift_avx: 80.1
vvc_h_loop_filter_chroma_8_mix_shift_c: 100.1
vvc_h_loop_filter_chroma_8_mix_shift_avx: 70.1
vvc_h_loop_filter_chroma_8_one-side_no-shift_c: 90.1
vvc_h_loop_filter_chroma_8_one-side_no-shift_avx: 70.1
vvc_h_loop_filter_chroma_8_one-side_shift_c: 139.8
vvc_h_loop_filter_chroma_8_one-side_shift_avx: 70.1
vvc_h_loop_filter_chroma_8_strong_no-shift_c: 80.1
vvc_h_loop_filter_chroma_8_strong_no-shift_avx: 70.1
vvc_h_loop_filter_chroma_8_strong_shift_c: 140.1
vvc_h_loop_filter_chroma_8_strong_shift_avx: 70.1
vvc_h_loop_filter_chroma_8_weak_no-shift_c: 60.1
vvc_h_loop_filter_chroma_8_weak_no-shift_avx: 20.1
vvc_h_loop_filter_chroma_8_weak_shift_c: 89.8
vvc_h_loop_filter_chroma_8_weak_shift_avx: 20.1
vvc_h_loop_filter_chroma_10_mix_no-shift_c: 59.8
vvc_h_loop_filter_chroma_10_mix_no-shift_avx: 30.1
vvc_h_loop_filter_chroma_10_mix_shift_c: 110.1
vvc_h_loop_filter_chroma_10_mix_shift_avx: 30.1
vvc_h_loop_filter_chroma_10_one-side_no-shift_c: 89.8
vvc_h_loop_filter_chroma_10_one-side_no-shift_avx: 30.1
vvc_h_loop_filter_chroma_10_one-side_shift_c: 140.1
vvc_h_loop_filter_chroma_10_one-side_shift_avx: 30.1
vvc_h_loop_filter_chroma_10_strong_no-shift_c: 90.1
vvc_h_loop_filter_chroma_10_strong_no-shift_avx: 30.1
vvc_h_loop_filter_chroma_10_strong_shift_c: 140.1
vvc_h_loop_filter_chroma_10_strong_shift_avx: 30.1
vvc_h_loop_filter_chroma_10_weak_no-shift_c: 60.1
vvc_h_loop_filter_chroma_10_weak_no-shift_avx: 30.1
vvc_h_loop_filter_chroma_10_weak_shift_c: 80.1
vvc_h_loop_filter_chroma_10_weak_shift_avx: 20.1
vvc_h_loop_filter_chroma_12_mix_no-shift_c: 70.1
vvc_h_loop_filter_chroma_12_mix_no-shift_avx: 29.8
vvc_h_loop_filter_chroma_12_mix_shift_c: 99.8
vvc_h_loop_filter_chroma_12_mix_shift_avx: 70.1
vvc_h_loop_filter_chroma_12_one-side_no-shift_c: 90.1
vvc_h_loop_filter_chroma_12_one-side_no-shift_avx: 70.1
vvc_h_loop_filter_chroma_12_one-side_shift_c: 130.1
vvc_h_loop_filter_chroma_12_one-side_shift_avx: 70.1
vvc_h_loop_filter_chroma_12_strong_no-shift_c: 90.1
vvc_h_loop_filter_chroma_12_strong_no-shift_avx: 70.1
vvc_h_loop_filter_chroma_12_strong_shift_c: 139.8
vvc_h_loop_filter_chroma_12_strong_shift_avx: 70.1
vvc_h_loop_filter_chroma_12_weak_no-shift_c: 50.1
vvc_h_loop_filter_chroma_12_weak_no-shift_avx: 30.1
vvc_h_loop_filter_chroma_12_weak_shift_c: 80.1
vvc_h_loop_filter_chroma_12_weak_shift_avx: 20.1
```

File: tests/checkasm/vvc_deblock.c [hunk 1]
```diff
@@ -72,59 +72,67 @@ static void randomize_params(int32_t beta[4], int32_t tc[4], const int size)
     }
 }
 
-static void randomize_chroma_buffers(int type, int beta[4], int32_t tc[4],
-   uint8_t *buf, ptrdiff_t xstride, ptrdiff_t ystride, const int shift, const int bit_depth)
+static void randomize_chroma_buffers(int type, int beta[4], int32_t tc[4], uint8_t *buf0, uint8_t *buf1,
```
Comment (nuomi2021) [inline tests/checkasm/vvc_deblock.c]: the type will indicate strong/weak/oneside

File: tests/checkasm/vvc_deblock.c [hunk 2]
```diff
@@ -134,22 +142,50 @@ static void check_deblock_chroma(const VVCDSPContext *h, int bit_depth)
...
 
-    if (check_func(h->lf.filter_chroma[0], "vvc_h_loop_filter_chroma_%d", bit_depth))
-    {
-
-        uint8_t *buf = buf0 + xstride * 5;
-
-        randomize_buffers(buf0, buf1, BUF_SIZE);
-        randomize_chroma_buffers(0, beta, tc, buf, xstride, ystride, shift, bit_depth);
-        memcpy(buf0, buf1, BUF_SIZE);
+    for(int j = 0; j < sizeof(types)/sizeof(char*); ++j) {
```
Comment (nuomi2021) [inline tests/checkasm/vvc_deblock.c]: this can be moved to randomize_chroma_buffers

**Comment by stone-d-chen** [comment]
Made the suggested changes and pushed.

I also changed the function signature so that the parameters important for testing different code branches are first for clarity.

File: tests/checkasm/vvc_deblock.c [hunk 3]
```diff
@@ -72,59 +72,96 @@ static void randomize_params(int32_t beta[4], int32_t tc[4], const int size)
...
+uint8_t* max_len_p, uint8_t* max_len_q, int beta[4], int32_t tc[4], 
+uint8_t *buf0, uint8_t *buf1, ptrdiff_t xstride, ptrdiff_t ystride)
 {
     const int size = shift ? 4 : 2;
     const int end = 8 / size;
 
+    switch (type) {
+        case 0: // strong
+            for(int i = 0; i < 4; ++i) {
+                max_len_p[i] = 3;
```
Comment (nuomi2021) [inline tests/checkasm/vvc_deblock.c:90]: the size of max_len_p is not always 4, if shift is zero, max_len_{p, q} is an array with 2 components. 
so you can move this to L188

File: tests/checkasm/vvc_deblock.c [hunk 4]
```diff
@@ -72,59 +72,96 @@ static void randomize_params(int32_t beta[4], int32_t tc[4], const int size)
     }
 }
 
-static void randomize_chroma_buffers(int type, int beta[4], int32_t tc[4],
-   uint8_t *buf, ptrdiff_t xstride, ptrdiff_t ystride, const int shift, const int bit_depth)
+static void randomize_chroma_buffers(const int bit_depth, const int type, const int shift, const int spatial_strong,
+uint8_t* max_len_p, uint8_t* max_len_q, int beta[4], int32_t tc[4],
```
Comment (nuomi2021) [inline tests/checkasm/vvc_deblock.c:80]: indent, better change to max_len_{p,q}[4] like beta 4.
better change beta to int32_t

File: tests/checkasm/vvc_deblock.c [hunk 5]
```diff
@@ -72,59 +72,96 @@ static void randomize_params(int32_t beta[4], int32_t tc[4], const int size)
...
-            SET(P1, RANDCLIP(P0, b3diff));
-            b3 -= b3diff = rnd() % FFMAX(b3 - (1 << (bit_depth - 8)), 1);
-            SET(Q1, RANDCLIP(Q0, b3diff));
-
-            buf += ystride;
+    randomize_buffers(buf0, buf1, BUF_SIZE);
+
+    uint8_t* buf = buf0 + xstride * 5;
+
+    if (spatial_strong) {
```
Comment (nuomi2021) [inline tests/checkasm/vvc_deblock.c:115]: why we not use the type to decide strong, weak, and one side strong?

File: tests/checkasm/vvc_deblock.c [hunk 6]
```diff
@@ -72,59 +72,96 @@ static void randomize_params(int32_t beta[4], int32_t tc[4], const int size)
...
+        for (int j = 0; j < size; j++) {
+            const int tc25 = TC25(j) << (bit_depth - 10);
```
Comment (nuomi2021) [inline tests/checkasm/vvc_deblock.c]: weak, strong, one side strong can be mixed.
maybe we need to add some commentslike hevc_deblock.c

File: tests/checkasm/vvc_deblock.c (hunk 4 above)
Comment (nuomi2021) [reply tests/checkasm/vvc_deblock.c:80]: type is the most important thing for the entire function, maybe we can move it to the first one

File: tests/checkasm/vvc_deblock.c (hunk 5 above)
Comment (stone-d-chen) [reply tests/checkasm/vvc_deblock.c:115]: It's possible to have max_len_{p,q} = 1 but the spatial activity calculation allow for strong. In this case the code will ignore the spatial activity and still use weak chroma deblocking. Vice versa, it's possible to have max_len_{p,q} = 3 but the spatial activity decide to go for weak. Essentially it allows for testing the spatial activity decision process from other elements.

**Comment by stone-d-chen** [comment]
Quick update:

- submitted a WIP branch for some perf optimizations, mainly so I didn't lose the train of thought for it; pausing now
- For some reason, the random TC / Beta values (ala HEVC) is not causing Strong spatial values to be generated, only weak; my original hacked version for testing strong/onesided used unrealistically large TC/Beta values to guarantee that strong would be calculated. **EDIT: This is fixed with the latest version, simple memcopy problem.**

File: tests/checkasm/vvc_deblock.c (hunk 5 above)
Comment (nuomi2021) [reply tests/checkasm/vvc_deblock.c:115]: How about this:
```c
enum {
    DEBLOCK_LONG
    DEBLOCK_STRONG
    DEBLOCK_ONE_SIDE
    DEBLOCK_WEAK
} DEBLOCK_TYPE

struct Testcase {
    DEBLOCK_TYPE type;
    uint8_t max_q;
    uint8_t max_q;
};
```
then we can generate an array of Testcase case[]

**Comment by nuomi2021** [comment]
> be calculated. **EDIT: This is fixed with the latest version, simple memcopy problem.**
you can send a pr for this. small patches are easy to review and merge.
thank you.

**Comment by stone-d-chen** [comment]
> you can send a pr for this. small patches are easy to review and merge.
> thank you.

Sounds good, do you mean send a patch to the mailing list?

**Comment by nuomi2021** [comment]
> > you can send a pr for this. small patches are easy to review and merge.
> > thank you.
> 
> Sounds good, do you mean send a patch to the mailing list?

to this repo, :)
still need some work to do before we send the patch upstream.

File: tests/checkasm/vvc_deblock.c (hunk 5 above)
Comment (stone-d-chen) [reply tests/checkasm/vvc_deblock.c:115]: The combination would be between types, and spatial strength, types are equivalent to max_len_{p,q}.
So it would be the following:

```c
enum {
    DEBLOCK_STRONG
    DEBLOCK_ONE_SIDE
    DEBLOCK_WEAK
} DEBLOCK_TYPE

enum {
    SPATIAL_STRONG
    SPATIAL_WEAK
} SPATIAL_ACTIVITY

struct Testcase {
    DEBLOCK_TYPE type;
    SPATIAL_ACTIVITY activity;
    max_len_{p,q}[4] // assuming this is an out variable
};
```

The intent is clearer but it might be better to hold off until more checkasm tests are done on creating the structs since we'd have a second loop to pack the test cases and the current one would unpack, I'm not sure it buys us anything atm. Also there's a few very specific edge cases I think would be good to introduce, so we might want an expanded struct.

**Comment by stone-d-chen** [comment]
@nuomi2021 this should be good to merge now

**Comment by nuomi2021** [comment]
thank you
